### ISSUE TITLE
resample function

### ISSUE THREAD

**Issue by simonster** [JuliaDSP/DSP.jl#104]
Now that we have @JayKickliter's excellent multirate work, it would be great to have a high-performance resampling function. I think this works to design an appropriate filter:

```julia
function resample_filter(ratio::Rational{Int}, transitionwidth, attenuation)
    # Determine resampling filter order
    L, beta = kaiserord(transitionwidth/num(ratio), attenuation)

    # Set cutoff so that we will attain `attenuation` dB rolloff
    cutoff = 1/max(num(ratio), den(ratio))-2*transitionwidth/num(ratio)

    # Design filter
    h = digitalfilter(Lowpass(cutoff), FIRWindow(kaiser(L, beta)))
    scale!(h, num(ratio))
end
```

The tricky part appears to be applying the filter such that the first output sample after ramp up corresponds to the first input sample (and the center of the filter).

**Comment by JayKickliter** [comment]
This definitely a good idea. I had some code to do it, but cut it out since it wasn't complete. I can resurrect it. My approach (from the user's perspective) was something like:

```julia
# Rational
myFilt = FIRFilter(3//4)
filt(myFilt, x)

# Arbitrary
myFilt = FIRFilter(3.13)
filt(myFilt, x)
```

The code you posted is still needed for creating the taps behind the scenes, but if all you want is to resample, I'd prefer to remove that step. 

That's a good point about ramp up. Some polyphase filtering code out there defaults to throwing away the ramp up portion. I believe the multi rate code I added is not technically setting the initial phase correctly for doing that. I'll look into it.

**Comment by simonster** [comment]
I agree that the API should be such that the user doesn't have to explicitly construct the filter, although it might make sense to leave the transition width and attenuation as optional arguments with sensible defaults.

I think it makes sense to provide both a convenient streaming API for resampling and a function like MATLAB `resample` that takes a whole signal and resamples it with zero phase delay, discarding the ramp up to the filter center (although SciPy `resample` uses an FFT which doesn't seem like a great idea for many reasons). It would be great if you could look into this; you're clearly the expert here.

**Comment by jcbsnd** [comment]
+1 A `resample` function with a simple API like the one in Matlab would be really great to have. Also, it would be nice to have some examples, like the ones in @JayKickliter's Multirate repository.

**Comment by JayKickliter** [comment]
I made a little progress over the weekend, but didn't have a whole lot of time.

Looking through the documentation I agree wee need more examples in general.

**Comment by JayKickliter** [comment]
![](https://farm9.staticflickr.com/8835/17389846312_523a21d130_o.png)
I added some [scratch code](https://github.com/JuliaDSP/DSP.jl/blob/jsk/resamplers/test/resample_filter.jl) to experiment with zero-ramp resampling. I'm at the same point I was last time I tried this. There's still glitch at the beginning. I don't know if it's unavoidable, or if I'm doing something wrong. Any ideas?

**Comment by jcbsnd** [comment]
If I understand correctly, Matlab zero-pads the signal to compensate for the filter's delay, which should eliminate that glitch in the beginning.

(Kind of related to the resample function, I created Julia versions of decimate (and cheby1), see https://gist.github.com/jcbsnd/1f54819708d4fcebc788)

**Comment by simonster** [comment]
It appears that there is compensation for the delay in the above plots, since the phases of the input and the output are the same. I think there will always be some kind of glitch at the beginning (and at the end, if you let the filter ramp down) although I have no idea what it should look like. I guess it's kind of surprising to me that the first sample of the output wouldn't match the first sample of the input, but I don't know whether this is something that comes from the filter design or the implementation.

@jcbsnd You would get better numerical accuracy simply by passing the output of `digitalfilter(Lowpass(wp), Chebyshev1(n, r))` to `filtfilt`, since it will use second order sections rather than computing a transfer function. Also, I think we want to use a polyphase FIR approach for decimation as well; it should be substantially faster because you don't have to filter the whole input signal.

**Comment by jcbsnd** [comment]
@simonster thanks for the advice, I did try to pass the digital filter directly to `filtfilt`, but found that this results in a significant error (numerical deviation) in the beginning of the decimated sample (at least for the examples I tested). I don't know why this is.

**Comment by JayKickliter** [comment]
I skimmed matlab's resample documentation. It mentions that 'edge effects' are to be expected when the first input sample is far away from zero. It makes sense I guess it makes sense, it's a big discontinuity. But in their example the first input/output samples match. They appear to only do rational resampling, which is different that the example I posted. I'll do a rational implementation to compare against theirs.

**Comment by jcbsnd** [comment]
I investigated the issue I mentioned above using the following example:

```julia
t = [0:0.0005:.1]
x = sin(2*pi*30*t) + sin(2*pi*60*t)
y = decimate(x,4)
```
First test, the coefficients (PolynomialRatio) are used as input to `filtfilt`:

![figure1](https://cloud.githubusercontent.com/assets/10650853/7524981/f7775c30-f505-11e4-8813-b8d997402d57.png)

Second test, the digital filter is used as input to `filtfilt`:

![figure2](https://cloud.githubusercontent.com/assets/10650853/7524989/077f77de-f506-11e4-9dd7-0b15896cf6db.png)

There is a significant error introduced when using the digital filter as input. Why this is, I don't know.

**Comment by simonster** [comment]
@jcbsnd It's possible that the way `filtfilt` extrapolates when using second order sections is not optimal. Right now we do extrapolation for each second order section, which is also what MATLAB does for `filtfilt` with SOS (and the results of our implementation and the MATLAB SOS implementation match according to our tests), but we may be better off doing it differently. I'll look into this.

@JayKickliter If you can do a rational implementation, that would be terrific. Then we could compare it directly against MATLAB with the same filter, which should be informative.

**Comment by JayKickliter** [comment]
I made some more progress, and added a rational `resample` function. The most surprising thing is that the output looks almost identical to the arbitrary version. Both have the same delay. Next step is to add some tests. @simonster, I may need your help generating some MATLAB data if I don't have it installed on any of my computers anymore. That will cover the rational resampler. I'm not yet sure how to test the other one.

**Comment by simonster** [comment]
@JayKickliter With the parameters in `resample_filter.jl`, I ran:

```julia
using MATLAB
matRat = mat"resample($x, $(float64(num(rerate))), $(float64(den(rerate))), $h)"
matRatTime = [0.0:length(matRat)-1]/float64(rerate)
@test_approx_eq matRat[1:length(yRat)] yRat
```

and the test passed, so I think it works! MATLAB also includes the ramp down in its output (running zeros through the filter) so that the output has length `ceil(Int, input*ratio)`. I think the `resample` function should probably do this on its own, but if I add the zeros to the end of the input, the result matches for the remaining points as well.

MATLAB seems to do something different for filter design. The docs say it uses `firls` (which we don't have) and a Kaiser window. It seems like the first output sample matches the first input sample, although there's definitely still a visible artifact at the beginning. But according to http://src.infinitewave.ca/, the quality of MATLAB `resample`'s antialiasing filter is not that great.

Given the similarity of the output between the rational version and the arbitrary version, I'd expect the arbitrary version is correct as well, although it would be great if there's a reference implementation we can compare it against.

**Comment by JayKickliter** [comment]
@simonster That's good news. I wish I still had MATLAB to do these tests, but I must admit I don't miss having to use it.

I figured out why the beginning looked so weird. I forgot (or failed to RTFM) that we are using normalized nyquist ƒ of 1 and not 0.5.

Despite being terribly useful, the arbitrary resampler is (quantifiably) inexact by nature. It's such a simple application of polyphase structures, but seems to be not well known and there few implementations of it. The only two I know of are GPL. I can compare it to the output of GNU Radio and/or Liquid. I don't know what else to use as a reference if that doesn't work.

Can expand on what meant by the ramp down? I read is as flushing the delay line, or `FIRFilter.history`, with `length(history)` 0's. Is that correct? I'll add it if that what MATLAB is doing.

**Comment by JayKickliter** [comment]
Never mind my last question. I see what you mean in the matlab documentation.

**Comment by JayKickliter** [comment]
I'd like to wrap this up this weekend. I found a couple bugs that prevent using `resample` with ratios that equate to pure interpolation/decimation. To fix them, I'll need some reference data.

@simonster: if possible, can you please run the following code through MATLAB and send the results?

```matlab
t = [0:100]
x = (1 + sin(2*pi*0.005*t)) .* sin(2*pi*.05*t)

[y1,b1] = resample(x, 1, 2)
[y2,b2] = resample(x, 2, 1)
[y3,b3] = resample(x, 3, 2)
[y4,b4] = resample(x, 2, 3)
```

**Comment by simonster** [comment]
Up at https://gist.github.com/simonster/9de556c0bf34c4a593be/raw/5a75eddb18429a5d2984ef413227d4cde754c1bf/results.mat

**Comment by JayKickliter** [comment]
Thanks

**Comment by JayKickliter** [comment]
All four cases are passing tests. I still need to add documentation. The tests are being run with with Matlab's taps, since we don't have `firls`.

![error](https://cloud.githubusercontent.com/assets/2551201/7672426/9bb82a76-fcaf-11e4-8c2c-b2b491f5a8cc.png)

In the figure below, you can see the response of Matlab's taps vs ours (created by `resample_filter`). The attenuation for ours (in red, forgot legend) is way too high. It's probably a stupid error (maybe β vs α), or frequency normalization.

![response](https://cloud.githubusercontent.com/assets/2551201/7672425/9ba10c24-fcaf-11e4-9cfb-c2acc764b967.png)

**Comment by simonster** [comment]
Fixed by @JayKickliter in #111

**Comment by Losses** [comment]
Any chance to make `decimate` in MATLAB available in Julia, but not only `resample`?
